### PR TITLE
[DOCS] Custom Actions

### PR DESCRIPTION
**Pull request by klavavej**
Resolves https://greatexpectations.atlassian.net/browse/DOC-900 according to the plan linked in that issue

Deploy Preview at https://deploy-preview-10772.docs.greatexpectations.io/docs/core/trigger_actions_based_on_results/create_a_custom_action 

- [X] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [X] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [X] Appropriate tests and docs have been updated

For more information about contributing, visit our [community resources](https://docs.greatexpectations.io/docs/core/introduction/community_resources#contribute-code-or-documentation).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: docs/docusaurus/docs/core/trigger_actions_based_on_results/create_a_checkpoint_with_actions.md [hunk 1]
```diff
@@ -11,14 +11,14 @@ import PrereqValidationDefinition from '../_core_components/prerequisites/_valid
...
-<h2>Prerequisites</h2>
+## Prerequisites
 
 - <PrereqPythonInstalled/>.
 - <PrereqGxInstalled/>.
 - <PrereqPreconfiguredDataContext/>. In this guide the variable `context` is assumed to contain your Data Context.
 - <PrereqValidationDefinition/>.
 
-### Procedure
+## Procedure
```
Comment (klavavej) [inline docs/docusaurus/docs/core/trigger_actions_based_on_results/create_a_checkpoint_with_actions.md:21]: **Note for reviewer:** apologies for scope creep but I noticed some header level issues on this page when I copied it to scaffold the new page, so I went ahead and fixed the problems on this page.

File: docs/docusaurus/docs/application_integration_support.md [hunk 2]
```diff
@@ -25,16 +25,17 @@ The following table defines the GX Cloud, GX Core, and Community Supported integ
...
 
 <sup>1</sup> We've also seen GX work with the following data sources in the past but we can't guarantee ongoing compatibility. These data sources include Clickhouse, Vertica, Dremio, Teradata, Athena, EMR Spark, AWS Glue, Microsoft Fabric, Trino, Pandas on (S3, GCS, Azure), Databricks (Spark), and Spark on (S3, GCS, Azure).<br/>
 <sup>2</sup> Support for BigQuery in GX Cloud will be available in a future release.<br/>
-<sup>3</sup> This includes configuration storage for Expectations, Checkpoints, Validation Definitions, and Validation Result<br/>
-<sup>4</sup> config_variables.yml<br/>
-<sup>5</sup> Although only Airflow is supported, GX Cloud and GX Core should work with any orchestrator that executes Python code.<br/>
-<sup>6</sup> Airflow version 2.9.0+ required<br/>
+<sup>3</sup> This includes configuration storage for Expectations, Checkpoints, Validation Definitions, and Validation Results.<br/>
+<sup>4</sup> We support the general workflow for creating custom Actions but cannot help troubleshoot the domain-specific logic within a custom Action.<br/>
+<sup>5</sup> Use `config_variables.yml`.<br/>
```
Comment (klavavej) [inline docs/docusaurus/docs/application_integration_support.md:36]: **Note for reviewer:** apologies for scope creep but I went ahead and made these footnotes more consistently structured as long as I was editing this content.

File: docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+"""
+This is an example script for how to create a custom Action.
```
Comment (klavavej) [inline docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py:2]: **Note for technical reviewer:** this is my first time putting a new code sample under test. I did get some errors with it that I was able to resolve so I think the test is working. But, please let me know if I need to add anything else to make sure this is sufficiently tested.

File: docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+pytest --docs-tests -k "docs_example_create_a_custom_action" tests/integration/test_script_runner.py
+"""
+
+# EXAMPLE SCRIPT STARTS HERE:
+
+# <snippet name="docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py - full code example">
+
+from typing import Literal
+
+from typing_extensions import override
```
Comment (klavavej) [inline docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py:14]: **Note for technical reviewer:** I got an error about override not being defined so I added this based on other code I see in the repo. It resolved the error so I think it's right but let me know if there's anything amiss here.

File: README.md [hunk 5]
```diff
@@ -70,7 +70,7 @@ To ensure the long-term quality of the GX Core codebase, we're not yet ready to
 | -------------------- | ------------------ | ----- |
 | CredentialStore      | 🟢 Ready           |       |
 | BatchDefinition      | 🟡 Partially ready | Formerly known as splitters |
-| Action               | 🔴 Not ready       |       |
+| Action               | 🟢 Ready           |       |
```
Comment (klavavej) [inline README.md:73]: **Note for product reviewer** Please confirm the contribution readiness status for Actions. I tried to ask about this in my docs plan but didn't get an answer so want to double check that I've got the right status.

File: great_expectations/checkpoint/actions.py [hunk 6]
```diff
@@ -182,10 +182,10 @@ def __new__(cls, clsname, bases, attrs):
 @public_api
 class ValidationAction(BaseModel, metaclass=MetaValidationAction):
     """
-    ValidationActions define a set of steps to be run after a validation result is produced.
+    Actions define a set of steps to run after a Validation Result is produced. Subclass `ValidationAction` to create a `custom Action </docs/core/trigger_actions_based_on_results/create_a_custom_action>`_.
```
Comment (klavavej) [inline great_expectations/checkpoint/actions.py]: **Note for reviewer**: though I've written "custom" in lowercase here, it's getting capitalized in the output. Is there any way I can force "custom" to display as lowercase here? this is my first time editing a docstring for our API reference
Screenshot 2024-12-13 at 1 10 03 PM

File: great_expectations/checkpoint/actions.py [hunk 7]
```diff
@@ -87,7 +87,7 @@ def _build_renderer(config: dict) -> Renderer:
         )
     return renderer
 
-
+@public_api
```
Comment (klavavej) [inline great_expectations/checkpoint/actions.py:91]: **Note for technical reviewer**: I added this to resolve an error about too many items missing the public_api decorator
Screenshot 2024-12-13 at 1 17 32 PM

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -308,7 +308,7 @@ jobs:
           # We decided to exclude all external HTTP requests but the ones that under the domain greatexpectations.io
           # The reason is to avoid having network errors such as pages that throw 429 after too many requests (like Github)
           # and to prevent other possible errors related to user agent or lychee capturing hrefs from metadata that don't resolve to a specific page (preconnects in JS)
-          args: "--exclude='http.*' --include='^https://(.+\\.)?greatexpectations\\.io/' 'docs/docusaurus/build/**/*.html'"
+          args: "--exclude='http.*' --include='docs/docusaurus/build/**/*.html'"
```
Comment (klavavej) [inline .github/workflows/ci.yml:311]: **Note for reviewers:** I'm temporarily changing the link checker config to workaround a [bug](https://greatexpectations.atlassian.net/browse/DOC-979) where the link checker fails when a PR contains a link to a page that that same PR is introducing.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -308,7 +308,7 @@ jobs:
           # We decided to exclude all external HTTP requests but the ones that under the domain greatexpectations.io
           # The reason is to avoid having network errors such as pages that throw 429 after too many requests (like Github)
           # and to prevent other possible errors related to user agent or lychee capturing hrefs from metadata that don't resolve to a specific page (preconnects in JS)
-          args: "--exclude='http.*' --include='^https://(.+\\.)?greatexpectations\\.io/' 'docs/docusaurus/build/**/*.html'"
+          args: "--exclude='http.*' 'docs/docusaurus/build/**/*.html'"
```
Comment (klavavej) [inline .github/workflows/ci.yml:309]: **Note for reviewers:** I'm temporarily changing the link checker config to work around a [bug](https://greatexpectations.atlassian.net/browse/DOC-979) where the link checker fails when a PR contains a link to a page that that same PR is introducing. I'll undo this change after merging this PR to restore link checking

Comment (cdkini) [reply .github/workflows/ci.yml:309]: Ah okay I have the same issue and didn't realize it was a documented bug - thanks for the call out!

File: README.md (hunk 5 above)
Comment (cdkini) [reply README.md:73]: Not the product review but LGTM from an engineering perspective - the API is stable and the steps to create custom actions are quite simple (as laid out here 🎉 )

File: docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py (hunk 3 above)
Comment (cdkini) [reply docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py:2]: Hmm so usually these snippet files represent actual tests and I don't think we have any test logic here.

At the bottom of the file, could we instantiate an instance of this custom class and assert against its run method?

```
my_action = MyCustomAction(...)
result = my_action.run(...)
assert result == {"some": "info"}
``` 

You may need to create a checkpoint and that can be a little cumbersome so let me know if you want me to jump in and add these details on your behalf.

File: docs/docusaurus/docs/core/trigger_actions_based_on_results/create_a_custom_action.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+---
+import TabItem from '@theme/TabItem';
+import Tabs from '@theme/Tabs';
+
+import PrereqPythonInstalled from '../_core_components/prerequisites/_python_installation.md';
+import PrereqGxInstalled from '../_core_components/prerequisites/_gx_installation.md';
+
+Great Expectations provides [Actions for common workflows](/application_integration_support.md#integrations) such as sending emails and updating Data Docs. If these don't meet your needs, you can create a custom Action to integrate with different tools or apply custom business logic based on Validation Results. Example use cases for custom Actions include:
+- Opening tickets in an issue tracker when Validation runs fail.
+- Sending emails to different teams depending on which Expectations fail.
```
Comment (cdkini) [inline docs/docusaurus/docs/core/trigger_actions_based_on_results/create_a_custom_action.md]: This has some overlap with our current email action right? Should we perhaps pick another example?

File: great_expectations/checkpoint/actions.py (hunk 7 above)
Comment (cdkini) [reply great_expectations/checkpoint/actions.py:91]: Ah amazing I was attempting to do the same thing here: https://github.com/great-expectations/great_expectations/pull/10752

Let's take in your change and then I'll update my PR to fill in any other gaps if we have them

File: great_expectations/checkpoint/actions.py (hunk 6 above)
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: Huh that is a little odd - I don't know the exact logic that does this but it might be ingrained in one of the docs libs we use.

Perhaps try MD syntax for hyperlinks? Not sure that's supported but worth a shot?

**Review by cdkini — CHANGES_REQUESTED** [review]
Looks really good just a minor point or two about the test file - happy to discuss just let me know!

File: docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py (hunk 3 above)
Comment (klavavej) [reply docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py:2]: Thank you - I would appreciate if you'd jump in and add the extra code to make this a sufficient test!

Comment (cdkini) [reply docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py:2]: I've updated the logic to show a concrete example of what a user might want to do but feel free to modify/critique as you see fit

Comment (klavavej) [reply docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py:2]: Thank you!

How does this custom calculated percentage of successful expectations across all validation results compare to the  `success_percent` info available through [`ExpectationSuiteValidationResult.statistics`](https://docs.greatexpectations.io/docs/reference/api/core/expectationsuitevalidationresult_class/)?

If these two things provide the same exact information in different ways, are there advantages to using a custom action instead of the built-in statistics?

Depending on how similar these two things are and what if any advantages there are to choosing one option over another in various scenarios, I think we should either acknowledge the similarities and tradeoffs or pick a different concrete example.

(this feedback is somewhat similar to the feedback you had for me about the email use case in the intro overlapping a bit with our built in email action.)

**Review by cdkini — APPROVED** [review]

File: docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py (hunk 3 above)
Comment (klavavej) [reply docs/docusaurus/docs/core/trigger_actions_based_on_results/_examples/create_a_custom_action.py:2]: connecting some dots: more discussion about this in slack at https://greatexpectationslabs.slack.com/archives/C05V0M18TEJ/p1734482967962709?thread_ts=1734360165.007409&cid=C05V0M18TEJ

File: great_expectations/checkpoint/actions.py (hunk 6 above)
Comment (klavavej) [reply great_expectations/checkpoint/actions.py]: MD syntax worked! 🎉
